Handle escape, unicode, and hex  in `tokenize_escaped_single_quoted_string`

JasonLi-cn opened the pull request:
Related to https://github.com/apache/arrow-datafusion/pull/9268

alamb reviewed (approved):
Thank you @JasonLi-cn  -- sorry for the delay in review. I think this PR does more than refactor the function, so I changed the name of the PR to reflect what I think is different

I also changed the API a little more to avoid some unecessary `pub (crate)`

Any chance you can take a look before I merge it to make sure I got it right?

JasonLi-cn commented:
> Thank you @JasonLi-cn -- sorry for the delay in review. I think this PR does more than refactor the function, so I changed the name of the PR to reflect what I think is different
> 
> I also changed the API a little more to avoid some unecessary `pub (crate)`
> 
> Any chance you can take a look before I merge it to make sure I got it right?

Thank you for reviewing this PR, and I think your modification is necessary.

JasonLi-cn commented on src/tokenizer.rs:
```diff
@@ -2139,4 +2236,68 @@ mod tests {
+        check_unescape(r"\xA9", None);
+
+        // octal byte value
+        check_unescape(r"\1", Some("\u{0001}"));
```
These existing tests have been compared with Pg and are consistent with Pg. As for the test for 'multi character strings', I can add some more.

alamb commented:
Thanks again @JasonLi-cn
